File: src/tools/customers.ts
import { JobBOSS2Client } from '../jobboss2-client.js';
import { ToolDefinition } from './tool-definition.js';
import {
    GetCustomersSchema,
    GetCustomerByIdSchema,
    CreateCustomerSchema,
    UpdateCustomerSchema,
} from '../schemas.js';

export const customerTools: ToolDefinition[] = [
    {
        name: 'get_customers',
        description: 'Retrieve a list of customers from JobBOSS2. Supports filtering, sorting, pagination, and field selection.',
    },
    {
        name: 'get_customer_by_code',
        description: 'Retrieve a specific customer by their customer code.',
    },
    {
        name: 'create_customer',
        description: 'Create a new customer in JobBOSS2.',
    },
    {
        name: 'update_customer',
        description: 'Update an existing customer in JobBOSS2.',
    },
];

export const customerHandlers: Record<string, (args: any, client: JobBOSS2Client) => Promise<any>> = {
    get_customers: async (args, client) => {
        const params = GetCustomersSchema.parse(args);
        return client.getCustomers(params);
    },
    get_customer_by_code: async (args, client) => {
        const { customerCode, fields } = GetCustomerByIdSchema.parse(args);
        return client.getCustomerById(customerCode, { fields });
    },
    create_customer: async (args, client) => {
        const customerData = CreateCustomerSchema.parse(args);
        return client.createCustomer(customerData);
    },
    update_customer: async (args, client) => {
        const { customerCode, ...updateData } = UpdateCustomerSchema.parse(args);
        return client.updateCustomer(customerCode, updateData);
    },
};

File: tools/customers.py
from typing import Any, Dict, List, Optional, Union
from fastmcp import FastMCP
from jobboss2_client import JobBOSS2Client

def register_customer_tools(mcp: FastMCP, client: JobBOSS2Client):
    @mcp.tool()
    async def get_customers(
        fields: str = None,
        sort: str = None,
        skip: int = None,
        take: int = 200,
        **kwargs
    ) -> List[Dict[str, Any]]:
        """Retrieve a list of customers from JobBOSS2. Supports filtering, sorting, pagination, and field selection."""
        params = {"fields": fields, "sort": sort, "skip": skip, "take": take, **kwargs}
        params = {k: v for k, v in params.items() if v is not None}
        return await client.get_customers(params)

    @mcp.tool()
    async def get_customer_by_code(customerCode: str, fields: str = None) -> Dict[str, Any]:
        """Retrieve a specific customer by their customer code."""
        params = {"fields": fields} if fields else None
        return await client.get_customer_by_code(customerCode, params)

    @mcp.tool()
    async def create_customer(
        customerCode: str,
        customerName: str,
        phone: str = None,
        billingAddress1: str = None,
        **kwargs
    ) -> Dict[str, Any]:
        """Create a new customer in JobBOSS2."""
        data = {
            "customerCode": customerCode,
            "customerName": customerName,
            "phone": phone,
            "billingAddress1": billingAddress1,
            **kwargs
        }
        data = {k: v for k, v in data.items() if v is not None}
        return await client.api_call("POST", "customers", data=data)

    @mcp.tool()
    async def update_customer(
        customerCode: str,
        customerName: str = None,
        phone: str = None,
        billingAddress1: str = None,
        **kwargs
    ) -> Dict[str, Any]:
        """Update an existing customer in JobBOSS2."""
        data = {
            "customerName": customerName,
            "phone": phone,
            "billingAddress1": billingAddress1,
            **kwargs
        }
        data = {k: v for k, v in data.items() if v is not None}
        return await client.api_call("PATCH", f"customers/{customerCode}", data=data)

